### PR TITLE
fix edge case that causes crash when unregistering

### PR DESCRIPTION
**Pull request by Elhodred**

### CODE REVIEW

File: src/mod/endpoints/mod_sofia/sofia_reg.c [hunk 1]
```diff
@@ -205,6 +205,16 @@ void sofia_reg_truly_del_gateway(sofia_profile_t *profile)
 				}
 			}
 
+			if (gateway_ptr->sofia_private) {
```
Comment (dev-ryanc) [inline src/mod/endpoints/mod_sofia/sofia_reg.c:208]: Move this on the statement below, it should be deleted during the state of `REG_STATE_DOWN`

File: src/mod/endpoints/mod_sofia/sofia_reg.c [hunk 2]
```diff
@@ -205,6 +205,16 @@ void sofia_reg_truly_del_gateway(sofia_profile_t *profile)
...
+				sofia_private_free(gateway_ptr->sofia_private);
+			}
+
+			if (gateway_ptr->nh) {
```
Comment (dev-ryanc) [inline src/mod/endpoints/mod_sofia/sofia_reg.c:212]: Move this on the statement below, it should be deleted during the state of REG_STATE_DOWN

**Review by dev-ryanc — CHANGES_REQUESTED** [review]

**Comment by dev-ryanc** [comment]
Make sure to run a sipp test on all possible scenario to make sure we don't introduce another crash

**Review by dev-ryanc — CHANGES_REQUESTED** [review]

File: src/mod/endpoints/mod_sofia/sofia.c [hunk 3]
```diff
@@ -1633,6 +1625,12 @@ static void our_sofia_event_callback(nua_event_t event,
 		}
 	}
 
+	if (event == nua_r_unregister && status != 401 && status != 407 && status >= 200) {
```
[the remark this answers is not in the document]
Comment (Elhodred) [reply src/mod/endpoints/mod_sofia/sofia.c:1628]: so I can send it to the done section safely and the clean up stuff runs. Is what avoided the lock that was happening

File: src/mod/endpoints/mod_sofia/sofia.c [hunk 4]
```diff
@@ -1633,6 +1625,14 @@ static void our_sofia_event_callback(nua_event_t event,
...
+		switch_log_printf(SWITCH_CHANNEL_LOG, SWITCH_LOG_DEBUG, "Mark gateway %s for destruction after unregister. Status %d.\n", sofia_private->gateway_name, status);
+		if (gateway) {
```
Comment (dev-ryanc) [inline src/mod/endpoints/mod_sofia/sofia.c:1630]: Why can't we put this on the if condition above? It doesn't do anything if we don't have the gateway ptr

**Review by dev-ryanc — CHANGES_REQUESTED** [review]

**Review by dev-ryanc — APPROVED** [review]

**Pull request merged**